keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1500 @@
+or other components besides kubectl.
+
+Syntax has always been a magnet for arguments in our industry.
+We expect this to be no exception.
+Reasonable people will disagree with our choices,
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+Given that KYAML is YAML it should be possible to revise
+without breaking anyone.
+
+### Risks and Mitigations
+
+Risk: Our encoder produces some output which is not valid YAML.
+Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.
+
+Risk: The result could be net worse than plain YAML.
+Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.
+
+Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
+Mitigation: Docs?
+
+Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
+Mitigation: Be as precise as possible where we are confident, but leave room where we are not.
+
+## Design Details
+
+Creating an encoder for this format is not very technically challenging,
+but there are many details to consider.
+
+### Feature gate
+
+Literal feature gates are not super useful for CLI features, so SIG CLI has a
+similar mechanism to gate features via environment variables.  We will use a
+variable named `"KUBECTL_KYAML"` to gate this feature.
+
+The output flags for JSON and YAML are handled in
+`k8s.io/cli-runtime/pkg/genericclioptions`.  Unfortunately, that package is
+already transitively depended on by `k8s.io/kubectl/pkg/cmd/util` where the CLI
+gates are defined. Rather than trying to break the import cycle, we will just
+check the environment variable directly, without the utils package.
+
+### Future compatibility
+
+KYAML is, and will remain, a strict subset of YAML.  Compliant YAML parsers
+should always be able to parse KYAML data and we will endeavour to make sure
+they can all retain fidelity with comments.
+
+It may be that, over time, we find some aspects of KYAML's exact rendering that
+can be inmproved, and we will consider those opportunities. The implication of
+this is that clients which rely on the byte-exact results of KYAML's rendering
+might be broken by future changes. We will try to avoid this, but we will not
+guarantee it.
+
+### Packaging
+
+We will create a new `kyaml` package under `sigs.k8s.io/yaml` for the rendering
+logic and put the printer under `staging/src/k8s.io/cli-runtime/pkg/printers`,
+using the renderer.
+
+### Components
+
+The primary component affected by this KEP is `kubectl`, but any other
+component which uses `k8s.io/cli-runtime/pkg/printers` and
+`k8s.io/cli-runtime/pkg/genericclioptions` will get KYAML support "for free".
```

Comment:
When we created that `cli-runtime ` package we tried to promote this heavily to be used by all the kubectl plugins, I believe a lot of them do use it, which means they will also support the new format, and by that we should get even broader feedback surface.